File: validating_clusters/.vdoc.py
# type: ignore
# flake8: noqa
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
import os
# local vs UGER
if os.path.expanduser('~') in ["/Users/youyun", "/Users/youyunzheng"]: 
    # in a local mac, the home directory is usuaully at '/Users/[username]'
    workdir = os.path.expanduser('~')+"/Documents/HMS/PhD/beroukhimlab/broad_mount/"
else:
    # in dipg or uger, the home directory is usuaully at '/home/unix/[username]'
    workdir = "/xchip/beroukhimlab/"

import pandas as pd
import numpy as np
import networkx as nx
# import torch_geometric
import sklearn
import matplotlib.pyplot as plt
from tqdm import tqdm
from joblib import Parallel, delayed
import multiprocessing
import pickle
import re
import scipy.sparse as sp
from sklearn.model_selection import StratifiedKFold
import umap.umap_ as umap
import csv
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
torch.manual_seed(123)
import numpy as np
np.random.seed(123)
import time
#
#
#
#
#

"""Adapted from https://github.com/weihua916/powerful-gnns/blob/master/util.py"""

class S2VGraph(object):
    def __init__(self, g, label, node_tags=None, node_features=None):
        '''
            g: a networkx graph
            label: an integer graph label
            node_tags: a list of integer node tags
            node_features: a torch float tensor, one-hot representation of the tag that is used as input to neural nets
            edge_mat: a torch long tensor, contain edge list, will be used to create torch sparse tensor
            neighbors: list of neighbors (without self-loop)
        '''
        self.label = label
        self.g = g
        self.node_tags = node_tags
        self.neighbors = []
        self.node_features = 0
        self.edge_mat = 0
        self.max_neighbor = 0

def create_s2v(graph_file):
    G = pickle.load(open(graph_file, 'rb'))
    graph_label = re.sub('_[0-9]*$','',list(G.nodes)[0])
    # relabel nodes to integer
    G = nx.relabel_nodes(G, {n: int(re.sub('.*_', '',n)) for n in G.nodes})
    # use fake label of 0 for everything
    return S2VGraph(G, label = 0), graph_label

def load_data_from_nx(graph_files):
    num_cores = multiprocessing.cpu_count()
    g_list, label_list = zip(*Parallel(n_jobs = num_cores)(
        delayed(create_s2v)(graph_file) for graph_file in graph_files if \
            # pickle.load(open(graph_file, 'rb')).number_of_nodes() > 9 and
            # pickle.load(open(graph_file, 'rb')).number_of_nodes() <= 30 and
            pickle.load(open(graph_file, 'rb')).number_of_edges() > 0
    ))
        
    return g_list, label_list
#
#
#
#
#
#
#
# graph_embeddings = pd.read_csv(workdir + "youyun/complexSV/data/TCGA/graph_embedding/UGformerV2_UnSup_lr0.0005_bs16_ep30_ff1024_nn5_sd512_do0.5_nl2_nt2_graph_embeddings.csv",index_col=0)
graph_embeddings = pd.read_csv(workdir + "youyun/complexSV/data/TCGA/graph_embedding/UGformerV2_UnSup_lr0.0005_bs16_ep60_ff1024_nn20_sd512_do0.5_nl3_nt2_graph_embeddings.csv",index_col=0)
ct_overlap = pd.read_csv(workdir + "youyun/complexSV/data/validation/cluster_overlaps_w_ct.csv")
#
#
#
#
#
graph_dir = workdir + "youyun/complexSV/data/TCGA/SV/nx/"
nx_files = [graph_dir+f+'_nx_graph.txt' for f in graph_embeddings.index]
print(len(nx_files))
graphs, labels = load_data_from_nx(nx_files)
#
#
#
#
#
# UMAP dimensionality reduction
reducer = umap.UMAP(
    n_neighbors = 50, min_dist=0.05, n_components=2, n_epochs = 500
)
embedding = reducer.fit_transform(graph_embeddings)
#
#
#
# Plotting
plt.figure(figsize=(10, 6))
plt.scatter(
    embedding[:, 0], embedding[:, 1], 
    c=[np.log10(graph.g.number_of_nodes()) for graph in graphs], 
    cmap='Spectral', s=5, alpha = 0.5
)
plt.title('UMAP Visualization of Graph Embeddings \n Colored by log10 SV Count')
plt.xlabel('UMAP 1')
plt.ylabel('UMAP 2')
plt.colorbar()
plt.show()
#
#
#
feature_names = [
    'svclass_DEL', 'svclass_DUP','svclass_TRA', 'svclass_h2hINV', 'svclass_t2tINV',
    'dist_to_line_breakend1','dist_to_sine_breakend1','dist_to_ltr_breakend1','dist_to_dna_breakend1',
    'dist_to_TAD_breakend1', 'dist_to_telomere_breakend1', 'dist_to_centromere_breakend1',
    'total_cn_breakend1', 'major_cn_breakend1', 'minor_cn_breakend1',
    'total_cn_breakend2', 'major_cn_breakend2', 'minor_cn_breakend2'
    
]
fig, axs = plt.subplots(
    math.ceil(len(feature_names) / 3), 3,
    figsize=(18, math.ceil(len(feature_names) / 2) * 2.5)
)
# need to adjust for the fact that there are three columns
for i, feat_name in enumerate(feature_names):
    axs[i//3,i % 3].scatter(
        embedding[:, 0], embedding[:, 1], 
        c=[
            sum(nx.get_node_attributes(graph.g, feat_name).values()) / graph.g.number_of_nodes()
#             np.std(list(nx.get_node_attributes(graph.g, feat_name).values()))
            for graph in graphs
        ], 
        cmap='Spectral', s=5, alpha = 0.5
    )
    axs[i//3,i % 3].set_title(feat_name)
# colorbar for each panel
fig.colorbar(
    plt.cm.ScalarMappable(
        norm=plt.Normalize(0, 1), cmap='Spectral'
    ), ax=axs, orientation='vertical',
    shrink = 1/math.ceil(len(feature_names) / 3)
)
plt.show()
#
#
#
# Plotting
plt.figure(figsize=(10, 6))
plt.scatter(
    embedding[:, 0], embedding[:, 1], 
    c=[ct_overlap.loc[(ct_overlap['cluster_ID'] == x), 'ct_overlap'] for x in graph_embeddings.index], 
    cmap='Spectral', s=5, alpha = 0.5
)
plt.title('UMAP Visualization of Graph Embeddings \n Colored by log10 SV Count')
plt.xlabel('UMAP 1')
plt.ylabel('UMAP 2')
plt.colorbar()
plt.show()
#
#
#
#
#
#
#
#
#

File: validating_clusters/.vdoc.r
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
library(knitr)
library(data.table)
library(ggplot2)
library(GenomicRanges)
# local vs UGER
if (Sys.getenv("HOME") %in% c("/Users/youyun", "/Users/youyunzheng")) {
    # in a local mac, the home directory is usuaully at '/Users/[username]'
    workdir <- "~/Documents/HMS/PhD/beroukhimlab/broad_mount/"
} else {
    # in dipg or uger, the home directory is usuaully at '/home/unix/[username]'
    workdir <- "/xchip/beroukhimlab/"
}
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
chromothripsis_overlap <- fread(paste0(workdir, "youyun/complexSV/data/validation/PCAWG_chromothripsisOverlap.txt"))
chromothripsis_overlap_pass_all = chromothripsis_overlap[(Pass_ALL)]
ct_pass_all_granges = GRanges(
    seqnames = gsub('23','X',gsub('^chr|p$|q$', '', chromothripsis_overlap_pass_all$chr)),
    ranges = IRanges(start = chromothripsis_overlap_pass_all$start, end = chromothripsis_overlap_pass_all$end),
    Sample = chromothripsis_overlap_pass_all$samplename
)
#
#
#
#
#
#
#
cluster_bedpe <- fread(paste0(workdir, "youyun/complexSV/data/TCGA/SV/total_bedpe/cluster_bedpe_concat.csv"))
cluster_ranges = rbind(melt(
    cluster_bedpe[,.(seqnames = seqnames_breakend1, start_breakend1, Sample, cluster_ID = paste0(Sample,'_',cluster_ID), SV_ID)],
    id.vars = c('Sample',"cluster_ID", "SV_ID",'seqnames'), variable.name = "breakend", value.name = "value"
),melt(
    cluster_bedpe[,.(seqnames = seqnames_breakend2, start_breakend2, Sample, cluster_ID = paste0(Sample,'_',cluster_ID), SV_ID)],
    id.vars = c('Sample',"cluster_ID", "SV_ID",'seqnames'), variable.name = "breakend", value.name = "value"
))[
    ,.(min = min(value), max = max(value)), by = c('Sample',"cluster_ID", "seqnames")
]
cluster_granges = GRanges(
    seqnames = cluster_ranges$seqnames,
    ranges = IRanges(start = cluster_ranges$min, end = cluster_ranges$max),
    cluster_ID = cluster_ranges$cluster_ID,
    Sample = cluster_ranges$Sample
)
#
#
#
#
#
graph_embedding <- fread(paste0(
    workdir, "youyun/complexSV/data/TCGA/graph_embedding/UGformerV2_UnSup_lr0.0005_bs16_ep30_ff1024_nn5_sd512_do0.5_nl2_nt2_graph_embeddings.csv"
))
dim(graph_embedding)
rsums = apply(graph_embedding, 1, function(x) sum(as.numeric(x[2:89])))
summary(rsums)
#
#
#
#
#
#
#
#| results: hold
cluster_overlaps_w_ct = cluster_ranges[
    ,.(ct_overlap = length(findOverlaps(
        cluster_granges[cluster_granges$cluster_ID == cluster_ID], 
        ct_pass_all_granges[ct_pass_all_granges$Sample == Sample]
    )) > 0),
    .(cluster_ID, Sample)
]
print('Breakdown of clusters that overlap with chromothripsis events:')
print(table(cluster_overlaps_w_ct$ct_overlap))
write.table(
    cluster_overlaps_w_ct,
    paste0(workdir, "youyun/complexSV/data/validation/cluster_overlaps_w_ct.csv"),
    sep = ",", quote = F, row.names = F

)
#
#
#
#
#
#
